readdir.go

```diff
@@ -0,0 +1,59 @@
+package hdfs
+
+import (
+	"context"
+	"io"
+
+	"github.com/beyondstorage/go-storage/v4/types"
+)
+
+const defaultListObjectLimit = 100
+
+type listDirInput struct {
+	rp  string
+	dir string
+
+	continuationToken string
+}
+
+func (i *listDirInput) ContinuationToken() string {
+	return i.continuationToken
+}
+
+func (s *Storage) listDirNext(ctx context.Context, page *types.ObjectPage) (err error) {
+	input := page.Status.(*listDirInput)
+
+	dir, err := s.hdfs.Open(input.rp)
+	if err != nil {
+		return err
+	}
+
+	for {
```

Comment:
Please don't use `for` in `listDirNext`. 

Our iterator will do this job: 
https://github.com/beyondstorage/go-storage/blob/de387baae5e6b246358c10ce01e333fb2d5f454b/types/iterator.generated.go#L125-L154

Our iterator provides `Next()` for users to call. And it will call `listDirNext` internally when all data has been consumed, so we don't need to do a loop by ourselves.

- Read `limit` files and update `page.Data`.
- Return `IterateDone` if there are no more files.